### PR TITLE
server: PATCH Endpoint Headers

### PR DESCRIPTION
**Pull request by jaymell**
## Motivation

Implement PATCH method for updating endpoint headers; update axum extension setup to remove deprecated APIs

## Solution

Merge new/modified headers with existing

### CODE REVIEW

File: server/svix-server/src/main.rs [hunk 1]
```diff
@@ -102,12 +102,12 @@ async fn main() {
     let mut app = Router::new()
         .nest("/api/v1", v1::router())
         .layer(TraceLayer::new_for_http().on_request(()))
-        .layer(AddExtensionLayer::new(pool.clone()))
-        .layer(AddExtensionLayer::new(queue_tx.clone()))
-        .layer(AddExtensionLayer::new(cfg.clone()));
+        .layer(Extension(pool.clone()))
+        .layer(Extension(queue_tx.clone()))
+        .layer(Extension(cfg.clone()));
```
Comment (tasn) [inline server/svix-server/src/main.rs:107]: Are you sure this is correct? (really deprecated)
If so, can you please:
1. Split it to its own PR.
2. Reference in the commit where you got this information from (e.g. a link to the docs).

File: server/svix-server/src/v1/endpoints/endpoint.rs [hunk 2]
```diff
@@ -208,11 +216,25 @@ impl From<EndpointHeaders> for EndpointHeadersOut {
...
+    #[validate]
+    headers: EndpointHeaders,
+}
+
+impl ModelIn for EndpointHeadersPatchIn {
     type ActiveModel = endpoint::ActiveModel;
 
     fn update_model(self, model: &mut Self::ActiveModel) {
-        model.headers = Set(Some(self.headers));
+        let existing = mem::replace(&mut model.headers, Set(None)).unwrap();
```
Comment (tasn) [inline server/svix-server/src/v1/endpoints/endpoint.rs:230]: Not gonna lie, I don't remember what `mem::replace` does, but it feels very wrong to use `mem`. What are you trying to do here?

File: server/svix-server/src/main.rs (hunk 1 above)
Comment (tasn) [reply server/svix-server/src/main.rs:107]: I did something with Axum now (testing a fix for a bug I reported) and this deprecation seems to only be done in the to-be-released version, and not in the currently released one. No?

Comment (jaymell) [reply server/svix-server/src/main.rs:107]: No problem, I will remove from PR for now. `AddExtensionLayer` is deprecated since 0.4.7, which was released a few days ago: https://github.com/tokio-rs/axum/releases/tag/axum-v0.4.7

Comment (tasn) [reply server/svix-server/src/main.rs:107]: Yeah, you're right, I dunno why it's not in the axum changelog. :|

https://github.com/tokio-rs/axum/blob/main/axum/CHANGELOG.md

Comment (tasn) [reply server/svix-server/src/main.rs:107]: Please open a PR for it, happy to insta-merge it, we don't want the warnings. :)

Comment (jaymell) [reply server/svix-server/src/main.rs:107]: In any event, I should have updated Cargo.toml as well to explicitly reference the latest version :|

Comment (tasn) [reply server/svix-server/src/main.rs:107]: Yeah, probably. :P 

Cargo "best practices" is not to commit the lock file for binaries, only libs (dunno whether we'll continue following this, but this is the recommendation I've seen). So I didn't get the update that showed it for you as deprecated. :) 

But yeah, just bump to 0.4.8 (I think that's the latest) and open a PR with this small change. Looking forward to having it in!

File: server/svix-server/src/v1/endpoints/endpoint.rs (hunk 2 above)
Comment (jaymell) [reply server/svix-server/src/v1/endpoints/endpoint.rs:230]: I agree it seems strange, but `mem::replace` is the lowest-overhead way I know of to modify a non-`Copy` field from a mutably borrowed struct: https://doc.rust-lang.org/error-index.html#E0507. 

The simplest alternative I know is to clone the field but was trying to avoid it.

Comment (tasn) [reply server/svix-server/src/v1/endpoints/endpoint.rs:230]: Yeah, definitely shouldn't copy the structure, but why wouldn't something like this work?
```rust
        model.headers = match model.headers {
            Some(&mut existing) => {
                existing.0.extend(self.headers.0);
                Set(Some(existing))
            }
            None => Set(Some(self.headers)),
        };
```
As in without that line? Why do you even need to touch model.headers before?

Comment (jaymell) [reply server/svix-server/src/v1/endpoints/endpoint.rs:230]: If I use something like above:
```rust
        model.headers = match model.headers {
            Some(existing) => {
                existing.0.extend(self.headers.0);
                Set(Some(existing))
            }
            None => Set(Some(self.headers)),
        };
```

I get the move error. Let me go back to the drawing board for a bit and see if there is a cleaner way.

Comment (tasn) [reply server/svix-server/src/v1/endpoints/endpoint.rs:230]: What are you unwrapping? Also, I think this is actually something that also changed in axum (though maybe in the unreleased version).

Comment (jaymell) [reply server/svix-server/src/v1/endpoints/endpoint.rs:230]: `model.headers` is a `sea_orm::ActiveValue` -- unwrapping if I want direct access to the underlying EndpointHeaders struct. I'm trying to see if I can do the pattern matching on the ActiveValue itself rather than the unwrapped value.

Comment (tasn) [reply server/svix-server/src/v1/endpoints/endpoint.rs:230]: Yeah, just do pattern matching there too, or maybe you can use `take()`.
```rust
        model.headers = match model.headers.take() {
            Some(existing) => {
                existing.0.extend(self.headers.0);
                Set(Some(existing))
            }
            None => Set(Some(self.headers)),
        };
```

File: server/svix-server/src/main.rs (hunk 1 above)
Comment (tasn) [reply server/svix-server/src/main.rs:107]: I can't unsee it now because of you. Let me know if you prefer me to open a PR and I will. I want it fixed, haha. :)

Comment (jaymell) [reply server/svix-server/src/main.rs:107]: Raised #283

Comment (tasn) [reply server/svix-server/src/main.rs:107]: Thanks!

File: server/svix-server/src/v1/endpoints/endpoint.rs (hunk 2 above)
Comment (jaymell) [reply server/svix-server/src/v1/endpoints/endpoint.rs]: Done.

**Review by tasn — APPROVED** [review]
